File: src/internal/adapter/broadcast/EventBroadcastAdapter.ts
import type { EventBroadcastAPI, PlayerEventMap, THEOplayer, StringKeyOf } from 'react-native-theoplayer';
import type { THEOplayerAdapter } from '../THEOplayerAdapter';
import { getNativeModule } from '../../utils/ModuleUtils';

const NativeEventBroadcastModule = getNativeModule('EventBroadcast');

export class EventBroadcastAdapter implements EventBroadcastAPI {
  constructor(private _player: THEOplayer) {}

  broadcastEvent<K extends StringKeyOf<PlayerEventMap>>(event: PlayerEventMap[K]): void {
    // Broadcast ReactNative event.
    (this._player as THEOplayerAdapter).dispatchEvent(event);

    try {
      // Broadcast native event.
      NativeEventBroadcastModule.broadcastEvent(this._player.nativeHandle, Object.freeze(event));
    } catch (e) {
      console.warn(`EventBroadcastModule not available: ${e}`);
    }
  }
}

File: android/src/main/java/com/theoplayer/broadcast/EventBroadcastAdapter.kt
package com.theoplayer.broadcast

import com.facebook.react.bridge.ReadableMap
import com.theoplayer.ReactTHEOplayerView

class EventBroadcastAdapter(private val view: ReactTHEOplayerView): DefaultEventDispatcher() {
  /**
   * Convert a react-native event to a native event and broadcast it.
   */
  fun broadcastEvent(event: ReadableMap) {
    EventAdapter.parseEvent(event)?.also {
      view.broadcast.dispatchEvent(it)
    }
  }
}
